File: backend/app/services/fitting/spin_system.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Sequence, Set, Tuple

AA_3TO1 = {
    "ALA": "A", "ARG": "R", "ASN": "N", "ASP": "D", "CYS": "C",
    "GLU": "E", "GLN": "Q", "GLY": "G", "HIS": "H", "ILE": "I",
    "LEU": "L", "LYS": "K", "MET": "M", "PHE": "F", "PRO": "P",
    "SER": "S", "THR": "T", "TRP": "W", "TYR": "Y", "VAL": "V",
}

AA_1TO3 = {v: k for k, v in AA_3TO1.items()}


@dataclass(frozen=True)
class SpinSystemKey:
    """
    Typed representation of an NMR spin-system key.
    Examples:
      - G2N: res_num=2, symbol='G', spins=('N',)
      - 2N: res_num=2, symbol='', spins=('N',)
      - G2N-HN: res_num=2, symbol='G', spins=('N', 'HN')
      - G2HN-N: res_num=2, symbol='G', spins=('HN', 'N')
      - L3CD1-HD1: res_num=3, symbol='L', spins=('CD1', 'HD1')
    """
    res_num: int
    symbol: str = ""
    spins: Tuple[str, ...] = ()
    raw: str = ""

    @classmethod
    def parse(cls, key_str: str) -> SpinSystemKey:
        if not key_str:
            return cls(res_num=0, raw="")

        raw_clean = key_str.strip()
        parts = raw_clean.split("-")

        # Parse first part (e.g. "GLY14N", "G14N", "14N", "40ASN", "71LYS", "L3CD1", "3CD1")
        p0 = parts[0].strip()
        m_num3 = re.match(r"^(\d+)([A-Za-z]{3})(.*)$", p0)
        m3 = re.match(r"^([A-Za-z]{3})(\d+)(.*)$", p0)
        m1 = re.match(r"^([A-Za-z]?)(\d+)(.*)$", p0)
        m_num1 = re.match(r"^(\d+)([A-Za-z])(.*)$", p0)

        if m_num3:
            rnum = int(m_num3.group(1))
            aa3 = m_num3.group(2).upper()
            sym = AA_3TO1.get(aa3, aa3[0])
            spin0 = m_num3.group(3).strip()
        elif m3:
            aa3 = m3.group(1).upper()
            sym = AA_3TO1.get(aa3, aa3[0])
            rnum = int(m3.group(2))
            spin0 = m3.group(3).strip()
        elif m1 and m1.group(2):
            sym = m1.group(1).upper()
            rnum = int(m1.group(2))
            spin0 = m1.group(3).strip()
        elif m_num1:
            rnum = int(m_num1.group(1))
            letter = m_num1.group(2).upper()
            rem = m_num1.group(3).strip()
            if letter in ("N", "C", "H"):
                sym = ""
                spin0 = f"{letter}{rem}"
            else:
                sym = letter
                spin0 = rem
        else:
            sym = ""
            rnum = 0
            spin0 = p0

        spins_list: List[str] = []
        if spin0:
            spins_list.append(spin0.upper())
        elif rnum > 0:
            spins_list.append("N")

        for p in parts[1:]:
            p_clean = p.strip()
            # If sub-spin has residue prefix e.g. "G2HN" in "G2N-G2HN", strip prefix
            sub_m = re.match(r"^[A-Za-z]?\d+(.*)$", p_clean)
            if sub_m and sub_m.group(1):
                spins_list.append(sub_m.group(1).strip().upper())
            else:
                spins_list.append(p_clean.upper())

        return cls(
            res_num=rnum,
            symbol=sym,
            spins=tuple(spins_list),
            raw=raw_clean,
        )

    def format(self, style: str = "canonical", include_symbol: bool = True) -> str:
        """
        Format spin key according to style:
          - canonical: G2N, G2N-HN, L3CD1-HD1
          - short: 2N, 2N-HN, 3CD1-HD1 (no AA symbol)
          - chemex_15n: 2N (or G2N if symbol requested)
        """
        prefix = f"{self.symbol}{self.res_num}" if (include_symbol and self.symbol) else f"{self.res_num}"
        if not self.spins:
            return prefix

        first_spin = self.spins[0]
        first_part = f"{prefix}{first_spin}"
        if len(self.spins) == 1:
            return first_part

        other_spins = "-".join(self.spins[1:])
        return f"{first_part}-{other_spins}"

    @property
    def canonical(self) -> str:
        return self.format(include_symbol=True)

    @property
    def short(self) -> str:
        return self.format(include_symbol=False)

    def matches(self, other: SpinSystemKey) -> bool:
        """Check if two keys represent the same residue entity."""
        if self.res_num != other.res_num:
            return False
        if self.symbol and other.symbol and self.symbol != other.symbol:
            return False
        return True


def sort_spin_keys(keys: Sequence[str]) -> List[str]:
    """Sort keys numerically by residue number, then alphabetically."""
    parsed_pairs = [(SpinSystemKey.parse(k), k) for k in keys]
    parsed_pairs.sort(key=lambda p: (p[0].res_num, p[0].symbol, p[0].spins, p[1]))
    return [p[1] for p in parsed_pairs]


def resolve_numeric_range(range_expr: str, available_keys: Sequence[str]) -> Tuple[List[str], List[str]]:
    """
    Resolve range expression (e.g. "13-15, 25, 40-44") against available keys.
    Returns (matched_keys, unrecognized_tokens).
    """
    tokens = re.split(r"[,;\s]+", range_expr.strip())
    tokens = [t for t in tokens if t]
    target_numbers: Set[int] = set()
    unrecognized: List[str] = []

    for token in tokens:
        if "-" in token:
            parts = token.split("-")
            if len(parts) == 2:
                try:
                    s_num = int(re.sub(r"\D", "", parts[0]))
                    e_num = int(re.sub(r"\D", "", parts[1]))
                    if s_num <= e_num:
                        for n in range(s_num, e_num + 1):
                            target_numbers.add(n)
                    else:
                        unrecognized.append(token)
                except ValueError:
                    unrecognized.append(token)
            else:
                unrecognized.append(token)
        else:
            try:
                num = int(re.sub(r"\D", "", token))
                target_numbers.add(num)
            except ValueError:
                unrecognized.append(token)

    parsed_keys = [(SpinSystemKey.parse(k), k) for k in available_keys]
    matched = [k for p, k in parsed_keys if p.res_num in target_numbers]
    return matched, unrecognized


def match_spin_key_sets(
    source_keys: Sequence[str],
    target_keys: Sequence[str]
) -> Dict[str, Any]:
    """
    Match two sets of spin-system keys of potentially different formats.
    Returns:
      {
        "matched": [{"source": s, "target": t, "res_num": n}, ...],
        "unmatched_source": [...],
        "unmatched_target": [...]
      }
    """
    src_parsed = {s: SpinSystemKey.parse(s) for s in source_keys}
    tgt_parsed = {t: SpinSystemKey.parse(t) for t in target_keys}

    matched: List[Dict[str, Any]] = []
    matched_src: Set[str] = set()
    matched_tgt: Set[str] = set()

    for s, s_key in src_parsed.items():
        for t, t_key in tgt_parsed.items():
            if t in matched_tgt:
                continue
            if s_key.matches(t_key):
                matched.append({
                    "source": s,
                    "target": t,
                    "res_num": s_key.res_num or t_key.res_num,
                })
                matched_src.add(s)
                matched_tgt.add(t)
                break

    unmatched_src = [s for s in source_keys if s not in matched_src]
    unmatched_tgt = [t for t in target_keys if t not in matched_tgt]

    return {
        "matched": matched,
        "unmatched_source": unmatched_src,
        "unmatched_target": unmatched_tgt,
    }

File: frontend/src/lib/spinSystem.ts
/**
 * Typed NMR Spin-System Key implementation.
 *
 * Handles parsing, formatting, sorting by residue number, numeric range
 * resolution, and cross-format matching for single-spin (15N/13C),
 * two-spin (15N-1H TROSY, 1H-15N IP/AP), and methyl (13C-1H MQ) keys.
 */

export const AA_3TO1: Record<string, string> = {
  ALA: "A", ARG: "R", ASN: "N", ASP: "D", CYS: "C",
  GLU: "E", GLN: "Q", GLY: "G", HIS: "H", ILE: "I",
  LEU: "L", LYS: "K", MET: "M", PHE: "F", PRO: "P",
  SER: "S", THR: "T", TRP: "W", TYR: "Y", VAL: "V",
};

export const AA_1TO3: Record<string, string> = Object.entries(AA_3TO1).reduce(
  (acc, [three, one]) => ({ ...acc, [one]: three }),
  {}
);

export class SpinSystemKey {
  readonly resNum: number;
  readonly symbol: string;
  readonly spins: readonly string[];
  readonly raw: string;

  constructor(resNum: number, symbol = "", spins: string[] = [], raw = "") {
    this.resNum = resNum;
    this.symbol = symbol.toUpperCase();
    this.spins = Object.freeze(spins.map(s => s.toUpperCase()));
    this.raw = raw;
  }

  static parse(keyStr: string): SpinSystemKey {
    if (!keyStr) return new SpinSystemKey(0, "", [], "");

    const rawClean = keyStr.trim();
    const parts = rawClean.split("-");

    const p0 = parts[0].trim();
    let sym = "";
    let rnum = 0;
    let spin0 = "";

    const mNum3 = p0.match(/^(\d+)([A-Za-z]{3})(.*)$/);
    const m3 = p0.match(/^([A-Za-z]{3})(\d+)(.*)$/);
    const m1 = p0.match(/^([A-Za-z]?)(\d+)(.*)$/);
    const mNum1 = p0.match(/^(\d+)([A-Za-z])(.*)$/);

    if (mNum3) {
      rnum = parseInt(mNum3[1], 10) || 0;
      const aa3 = mNum3[2].toUpperCase();
      sym = AA_3TO1[aa3] || aa3[0];
      spin0 = mNum3[3].trim();
    } else if (m3) {
      const aa3 = m3[1].toUpperCase();
      sym = AA_3TO1[aa3] || aa3[0];
      rnum = parseInt(m3[2], 10) || 0;
      spin0 = m3[3].trim();
    } else if (m1 && m1[2]) {
      sym = m1[1].toUpperCase();
      rnum = parseInt(m1[2], 10) || 0;
      spin0 = m1[3].trim();
    } else if (mNum1) {
      rnum = parseInt(mNum1[1], 10) || 0;
      const letter = mNum1[2].toUpperCase();
      const rem = mNum1[3].trim();
      if (letter === "N" || letter === "C" || letter === "H") {
        sym = "";
        spin0 = `${letter}${rem}`;
      } else {
        sym = letter;
        spin0 = rem;
      }
    } else {
      sym = "";
      rnum = 0;
      spin0 = p0;
    }

    const spinsList: string[] = [];
    if (spin0) {
      spinsList.push(spin0.toUpperCase());
    } else if (rnum > 0) {
      spinsList.push("N");
    }

    for (let i = 1; i < parts.length; i++) {
      const p = parts[i].trim();
      // If sub-spin has residue prefix e.g. "G2HN" in "G2N-G2HN", strip prefix
      const subM = p.match(/^[A-Za-z]?\d+(.*)$/);
      if (subM && subM[1]) {
        spinsList.push(subM[1].trim().toUpperCase());
      } else {
        spinsList.push(p.toUpperCase());
      }
    }

    return new SpinSystemKey(rnum, sym, spinsList, rawClean);
  }

  format(style: "canonical" | "short" = "canonical"): string {
    const includeSymbol = style === "canonical";
    const prefix = includeSymbol && this.symbol ? `${this.symbol}${this.resNum}` : `${this.resNum}`;
    if (this.spins.length === 0) return prefix;

    const firstSpin = this.spins[0];
    const firstPart = `${prefix}${firstSpin}`;
    if (this.spins.length === 1) return firstPart;

    const otherSpins = this.spins.slice(1).join("-");
    return `${firstPart}-${otherSpins}`;
  }

  get canonical(): string {
    return this.format("canonical");
  }

  get short(): string {
    return this.format("short");
  }

  matches(other: SpinSystemKey): boolean {
    if (this.resNum !== other.resNum) return false;
    if (this.symbol && other.symbol && this.symbol !== other.symbol) return false;
    return true;
  }
}

/**
 * Parses any residue/spin key string into a typed SpinSystemKey.
 */
export function parseSpinKey(rawKey: string): SpinSystemKey {
  return SpinSystemKey.parse(rawKey);
}

/**
 * Normalizes a residue label to canonical form (e.g., "GLY13N" -> "G13N", "13N" -> "13N").
 */
export function normalizeResidueKey(label: string): string {
  if (!label) return "";
  const parsed = SpinSystemKey.parse(label);
  return parsed.canonical;
}

/**
 * Extracts a numeric value for sorting residues (e.g., "13N" -> 13, "G14N" -> 14, "L3CD1-HD1" -> 3).
 */
export function extractResidueNumber(resKey: string): number {
  return SpinSystemKey.parse(resKey).resNum;
}

/**
 * Sorts keys numerically by residue number regardless of prefix or format.
 */
export function sortSpinKeys(keys: string[]): string[] {
  return [...keys].sort((a, b) => {
    const pA = SpinSystemKey.parse(a);
    const pB = SpinSystemKey.parse(b);
    if (pA.resNum !== pB.resNum) {
      return pA.resNum - pB.resNum;
    }
    if (pA.symbol !== pB.symbol) {
      return pA.symbol.localeCompare(pB.symbol);
    }
    return a.localeCompare(b);
  });
}

/**
 * Deduplicates spin keys representing the same spin system entity (e.g. "32N" and "32N-H").
 * Prefers shorter keys (e.g. "32N" over "32N-H") or keys without trailing coupled partner.
 */
export function deduplicateSpinKeys(keys: string[]): string[] {
  const groups = new Map<string, string[]>();
  for (const k of keys) {
    const parsed = SpinSystemKey.parse(k);
    const gid = parsed.resNum > 0
      ? `${parsed.resNum}_${parsed.symbol}_${parsed.spins[0] || ''}`
      : `raw_${k}`;
    const list = groups.get(gid) || [];
    list.push(k);
    groups.set(gid, list);
  }

  const result: string[] = [];
  for (const list of groups.values()) {
    list.sort((a, b) => a.length - b.length || a.localeCompare(b));
    result.push(list[0]);
  }
  return sortSpinKeys(result);
}

/**
 * Resolves a range expression (e.g. "13-15, 25, 40-44") against available keys.
 */
export function resolveNumericRange(
  rangeExpr: string,
  availableKeys: string[]
): { matched: string[]; unmatched: string[] } {
  if (!rangeExpr || !rangeExpr.trim()) {
    return { matched: [], unmatched: [] };
  }

  const tokens = rangeExpr.trim().split(/[,;\s]+/).filter(Boolean);
  const targetNumbers = new Set<number>();
  const unmatched: string[] = [];

  for (const token of tokens) {
    if (token.includes("-")) {
      const parts = token.split("-");
      if (parts.length === 2) {
        const sNum = parseInt(parts[0].replace(/\D/g, ""), 10);
        const eNum = parseInt(parts[1].replace(/\D/g, ""), 10);
        if (!isNaN(sNum) && !isNaN(eNum) && sNum <= eNum) {
          for (let i = sNum; i <= eNum; i++) {
            targetNumbers.add(i);
          }
        } else {
          unmatched.push(token);
        }
      } else {
        unmatched.push(token);
      }
    } else {
      const num = parseInt(token.replace(/\D/g, ""), 10);
      if (!isNaN(num)) {
        targetNumbers.add(num);
      } else {
        unmatched.push(token);
      }
    }
  }

  const matched = availableKeys.filter(k => {
    const p = SpinSystemKey.parse(k);
    return targetNumbers.has(p.resNum);
  });

  return { matched, unmatched };
}

export interface SpinKeyMatchResult {
  matched: Array<{ source: string; target: string; resNum: number }>;
  unmatchedSource: string[];
  unmatchedTarget: string[];
}

/**
 * Matches two sets of spin-system keys of potentially different formats.
 */
export function matchSpinKeySets(
  sourceKeys: string[],
  targetKeys: string[]
): SpinKeyMatchResult {
  const srcParsed = sourceKeys.map(s => ({ key: s, parsed: SpinSystemKey.parse(s) }));
  const tgtParsed = targetKeys.map(t => ({ key: t, parsed: SpinSystemKey.parse(t) }));

  const matched: Array<{ source: string; target: string; resNum: number }> = [];
  const matchedSrc = new Set<string>();
  const matchedTgt = new Set<string>();

  for (const s of srcParsed) {
    for (const t of tgtParsed) {
      if (matchedTgt.has(t.key)) continue;
      if (s.parsed.matches(t.parsed)) {
        matched.push({
          source: s.key,
          target: t.key,
          resNum: s.parsed.resNum || t.parsed.resNum,
        });
        matchedSrc.add(s.key);
        matchedTgt.add(t.key);
        break;
      }
    }
  }

  const unmatchedSource = sourceKeys.filter(s => !matchedSrc.has(s));
  const unmatchedTarget = targetKeys.filter(t => !matchedTgt.has(t));

  return {
    matched,
    unmatchedSource,
    unmatchedTarget,
  };
}
